Design/Workbench/design-projectsaving.md

```diff
@@ -0,0 +1,229 @@
+
+To decode a normal type it would be a simple call to json.JSONDecoder which will utilise the following table for translating back to python objects:
+
+![JSON Decoding Table](./json-decoding-table.png) "*(Docs.python.org, 2018)*"
+
+With these recreated key-value pairs the whole point of this is to recreate the interfaces and display them to the user, this will be handled partially by the middle-man "Loader" which will take each object recreated by the ProjectReader and find it's relevant decoder. This will be achieved by asking the DecoderFactory for the correct Decoder this will be done by passing a string that represents the name loaded in from the file, this name should be present in one of the Decoder's tag list (Very similar to the encoder tag list and should be the same value actual implementation is undefined). When the DecoderFactory finds the tag it is looking for it will pass the Decoder back to the "Loader"
+
+Then the Loader will pass to the decoder the key-value pair object of state that is required to recreate the object and restore it's state. The best way to show this even though it will largely be left up to interface implementers to do this, is with an example:
+```python
+class exampleDecoder():
+  def decode(self, dict):
+    TestUI(dict["arg1"], dict["args2"])
+    TestUI.ui.spinBox.setValue(dict["spinBox"]) 
+    TestUI.ui.lineEdit.setValue(dict["lineEdit"]) 
+    TestUI.ui.dateEdit.setValue(dict["dateEdit"]) 
+    TestUI.ui.radioButton.setValue(dict["radioButton"]) 
+    TestUI.ui.checkBox.setValue(dict["checkBox"])
+    TestUI.ui.label_7.setVisible(dict["label_7Visible"])
+    TestUI.show()
+```
+
+Then this leaves custom Decoding from JSON to be tackled, a custom decoder is a little harder, than an encoder, as it relies on the key name when decoding being something unique to that type so would need to be on a type requirement by requirement basis. So in the JSON a programmer would need to add a _object_type_ or some other unique string to show that a type like that is there, but it's really on a per type basis. For example: 
+```python
+# (Docs.python.org, 2018)
+def object_type(dict):
+  # __object_type__ being a key that is unique to that type
+  if "__object_type__" in dict:
+    # Return an object of the original object type
+    return object(dict["object_part1"], dict["object_part2"], dict["object_part3"])
+  return dict
+```
+
+**C++**
+Similarly with the C++ area of serialisation exposing most of the required details to python will be the hardest part, however if most of the interface's details are exposed properly then the decoder and encoders should be able to work exactly the same on these interfaces if only with a little extra implementation on the UI porter/maintainer's part.
+
+Options for Implementation of Workspace Saving
```

Comment:
I think this is duplicated from the section above?

Reply:
Not quite, but I will expand the title of this subsection to explain a little better what it is